src/backend/joanie/payment/models.py

```diff
@@ -1 +1,463 @@
+from . import get_payment_backend
+
+User = get_user_model()
+
+
+class Invoice(models.Model):
+    """
+    Invoice model is an accounting element related to an order
+    """
+
+    parent = models.ForeignKey(
+        "self",
+        on_delete=models.RESTRICT,
+        related_name="children",
+        verbose_name="parent",
+        blank=True,
+        null=True,
+    )
+    reference = models.CharField(
+        db_index=True,
+        editable=False,
+        max_length=40,
+        unique=True,
+    )
+    created_on = models.DateTimeField(
+        _("created on"),
+        auto_now_add=True,
+        editable=False,
+    )
+    updated_on = models.DateTimeField(
+        _("updated on"),
+        auto_now=True,
+        editable=False,
+    )
+    order = models.OneToOneField(
+        Order,
+        verbose_name=_("order"),
+        on_delete=models.PROTECT,
+        blank=True,
+        null=True,
+    )
+    amount = MoneyField(
+        _("amount"),
+        max_digits=9,
+        decimal_places=2,
+        default_currency=settings.DEFAULT_CURRENCY,
+        null=False,
+        blank=False,
+    )
+    recipient_name = models.CharField(
+        _("invoice recipient"), max_length=40, null=False, blank=False
+    )
+    recipient_address = models.TextField(
+        _("invoice address"), max_length=255, null=False, blank=False
+    )
+    localized_context = models.JSONField(
+        _("context"),
+        help_text=_("Localized data that needs to be frozen on invoice creation"),
+        editable=False,
+    )
+
+    class Meta:
+        db_table = "joanie_invoice"
+        verbose_name = _("Invoice")
+        verbose_name_plural = _("Invoices")
+        constraints = [
+            models.CheckConstraint(
+                check=(models.Q(order__isnull=False) & models.Q(parent__isnull=True))
+                | (models.Q(order__isnull=True) & models.Q(parent__isnull=False)),
+                name="invoice_relies_on_order_or_invoice",
+            ),
+        ]
+
+    def __str__(self):
+        types = dict(payment_enums.INVOICE_TYPES)
+        return f"{types[self.type]} {self.reference}"
+
+    @property
+    def root_parent(self):
+        """
+        Return the root parent invoice.
+        """
+        parent = self
+
+        while parent.parent is not None:
+            parent = parent.parent
+
+        return parent
+
+    @property
+    def state(self):
+        """
+        Process the state of the invoice
+        """
+        if self.balance.amount >= 0:
+            if (
+                self.invoiced_balance.amount == 0
+                and self.transactions_balance.amount == 0
+            ):
+                return payment_enums.INVOICE_STATE_REFUNDED
+
+            return payment_enums.INVOICE_STATE_PAID
+
+        return payment_enums.INVOICE_STATE_UNPAID
+
+    def get_related_order(self):
+        """
+        Return the related order.
+        """
+        return self.root_parent.order
```

Comment:
maybe this is enough with the right db constraints:
```suggestion
        return self.order or self.parent.order
```